src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -244,16 +218,75 @@ Status RedisStoreClient::AsyncMultiGet(
   RAY_CHECK(callback);
   if (keys.empty()) {
     callback({});
+    return Status::OK();
   }
   std::vector<std::string> true_keys;
   for (auto &key : keys) {
     true_keys.push_back(GenRedisKey(external_storage_namespace_, table_name, key));
   }
-  RAY_CHECK_OK(MGetValues(
-      redis_client_, external_storage_namespace_, table_name, true_keys, callback));
+  MGetValues(table_name, true_keys, callback);
   return Status::OK();
 }
 
+void RedisStoreClient::SendRedisCmd(std::vector<std::string> keys,
+                                    std::vector<std::string> args,
+                                    RedisCallback redis_callback) {
+  RAY_CHECK(!keys.empty());
+  std::string debug_redis_cmds;
+  auto counter = std::make_shared<std::atomic<size_t>>(0);
+  auto send_redis = [this,
+                     debug_redis_cmds,
+                     counter = counter,
+                     keys,
+                     args = std::move(args),
+                     redis_callback = std::move(redis_callback)]() mutable {
+    *counter += 1;
+    // There are still pending requets for these keys.
+    if (*counter != keys.size()) {
+      return;
+    }
+
+    // Send the actual request
+    auto cxt = redis_client_->GetShardContext("");
+    RAY_CHECK_OK(
+        cxt->RunArgvAsync(std::move(args),
+                          [this,
+                           debug_redis_cmds,
+                           keys = std::move(keys),
+                           redis_callback = std::move(redis_callback)](auto reply) {
+                            std::vector<std::function<void()>> requests;
+                            // No new queues should have been added here
+                            RAY_CHECK(Progress(keys, [&requests](auto &queue) mutable {
```

Comment:
i think if we create two helper functions for Progress, one is 
PopRedisCommandByKeys, another is PushRedisCommandByKeys, should increase the code readability.